# This is a BitKeeper generated diff -Nru style patch.
#
# ChangeSet
#   2005/02/13 14:34:55+01:00 spock@spock.one.pl 
#   Fixed:
#   - displaying an empty char when fbcon is blanked
#   - oops when using fb drivers as modules
# 
# drivers/video/fbsplash.c
#   2005/02/13 14:34:42+01:00 spock@spock.one.pl +6 -1
#   Fixed a possible source of oopses which could occur when using fbcon compiled into the kernel and fb drivers as modules.
# 
# drivers/video/cfbsplash.c
#   2005/02/13 14:34:42+01:00 spock@spock.one.pl +2 -1
#   Fixed a little bug regarding displaying an empty character when the console is blanked.
# 
# ChangeSet
#   2005/02/13 13:31:05+01:00 spock@spock.one.pl 
#   Merge spock.one.pl:/devel/kernel/linux-2.5
#   into spock.one.pl:/devel/kernel/linux-fbsplash
# 
# include/linux/sysctl.h
#   2005/02/13 13:30:57+01:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# ChangeSet
#   2005/02/05 14:46:20+01:00 spock@spock.one.pl 
#   Merge spock.one.pl:/devel/kernel/linux-2.5
#   into spock.one.pl:/devel/kernel/linux-fbsplash
# 
# kernel/sysctl.c
#   2005/02/05 14:46:12+01:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# include/linux/sysctl.h
#   2005/02/05 14:46:12+01:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# drivers/video/Makefile
#   2005/02/05 14:46:12+01:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# drivers/video/Kconfig
#   2005/02/05 14:46:12+01:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# drivers/char/n_tty.c
#   2005/02/05 14:46:12+01:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# ChangeSet
#   2005/01/29 13:43:13+01:00 spock@spock.one.pl 
#   Merge spock.one.pl:/devel/kernel/linux-2.5
#   into spock.one.pl:/devel/kernel/linux-fbsplash
# 
# drivers/video/console/fbcon.c
#   2005/01/29 13:43:07+01:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# drivers/video/Kconfig
#   2005/01/29 13:43:06+01:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# ChangeSet
#   2005/01/24 01:18:29+01:00 spock@spock.one.pl 
#   Merge spock.one.pl:/devel/kernel/linux-2.5
#   into spock.one.pl:/devel/kernel/linux-fbsplash
# 
# include/linux/fb.h
#   2005/01/24 01:18:20+01:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# ChangeSet
#   2005/01/20 19:58:36+01:00 spock@spock.one.pl 
#   Shame on me :) It seems I put a chunk of fbsplash code into a different function
#   when merging some upstream changes. Moved it back where it belongs :)
# 
# drivers/video/console/fbcon.c
#   2005/01/20 19:58:26+01:00 spock@spock.one.pl +10 -12
#   Shame on me :) It seems I put a chunk of fbsplash code into a different function
#   when merging some upstream changes. Moved it back where it belongs :)
# 
# ChangeSet
#   2005/01/20 18:32:17+01:00 spock@spock.one.pl 
#   A fix for Gentoo bug #78210. It turns out that the colors are completely 
#   broken when using applications which save the current cmap on start and
#   restore it later on. This happens because the cmap created by fbsplash is
#   not available in the userspace. FBIOGETCMAP returns the cmap from 
#   info->cmap. Either I am missing something here, or the current cmap is
#   not stored in info->cmap (despite the comments in fb.h). I modified
#   fb_set_cmap() to save all cmap changes in info->cmap. If this is a bug
#   in the kernel, fb_set_cmap_user() will probably have to be modified as
#   well. 
#   
#   One additional modification was necessary as well. Calling setcolreg()
#   messes up the info->pseudo_palette (which works a little differently
#   when fbsplash is enabled). The new fbsplash_fix_pseudo_pal() is provided
#   to fix the problem. The function has to be called after all cmap changes.
# 
# drivers/video/fbsplash.h
#   2005/01/20 18:31:52+01:00 spock@spock.one.pl +4 -0
#   Added fbsplash_fix_pseudo_pal().
# 
# drivers/video/fbcmap.c
#   2005/01/20 18:31:52+01:00 spock@spock.one.pl +8 -0
#   Added fbsplash_fix_pseudo_pal() and cmap saving for in-kernel set_cmap().
# 
# drivers/video/console/fbcon.c
#   2005/01/20 18:31:51+01:00 spock@spock.one.pl +1 -15
#   Added fbsplash_fix_pseudo_pal().
# 
# drivers/video/cfbsplash.c
#   2005/01/20 18:31:51+01:00 spock@spock.one.pl +21 -0
#   Added fbsplash_fix_pseudo_pal().
# 
# ChangeSet
#   2005/01/19 21:14:33+01:00 spock@spock.one.pl 
#   Merged with upstream.
# 
# drivers/video/console/fbcon.c
#   2005/01/19 21:14:19+01:00 spock@spock.one.pl +3 -1
#   Merged with upstream.
# 
# drivers/video/Makefile
#   2005/01/19 21:14:19+01:00 spock@spock.one.pl +0 -0
#   Merged with upstream.
# 
# drivers/video/Kconfig
#   2005/01/19 21:09:24+01:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# drivers/char/keyboard.c
#   2005/01/19 21:09:24+01:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# drivers/Makefile
#   2005/01/19 21:09:24+01:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# ChangeSet
#   2005/01/15 14:28:34+01:00 spock@spock.one.pl 
#   Removed reference to fbsplash_putc() since it is no longer used.
# 
# drivers/video/fbsplash.h
#   2005/01/12 20:17:10+01:00 spock@spock.one.pl +0 -1
#   Removed reference to fbsplash_putc() since it is no longer used.
# 
# ChangeSet
#   2005/01/12 20:10:13+01:00 spock@spock.one.pl 
#   Merge
# 
# include/linux/console_struct.h
#   2005/01/12 20:10:10+01:00 spock@spock.one.pl +0 -0
#   SCCS merged
# 
# drivers/char/n_tty.c
#   2005/01/12 20:01:56+01:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# drivers/char/keyboard.c
#   2005/01/12 20:01:56+01:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# drivers/Makefile
#   2005/01/12 20:01:56+01:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# ChangeSet
#   2005/01/05 14:36:04+01:00 spock@spock.one.pl 
#   Merged upstream changes.
# 
# include/linux/sysctl.h
#   2005/01/05 14:35:52+01:00 spock@spock.one.pl +1 -1
#   Merged upstream changes.
# 
# drivers/video/console/fbcon.c
#   2005/01/05 14:35:52+01:00 spock@spock.one.pl +6 -9
#   Merged upstream changes.
# 
# kernel/sysctl.c
#   2005/01/05 14:28:15+01:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# include/linux/fb.h
#   2005/01/05 14:28:15+01:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# drivers/char/keyboard.c
#   2005/01/05 14:28:15+01:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# drivers/Makefile
#   2005/01/05 14:28:15+01:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# ChangeSet
#   2004/12/26 16:16:41+01:00 spock@spock.one.pl 
#   Removed fbsplash_putc since it's no longer needed for anything.
# 
# drivers/video/cfbsplash.c
#   2004/12/26 16:16:27+01:00 spock@spock.one.pl +0 -16
#   Removed fbsplash_putc since it's no longer needed for anything.
# 
# ChangeSet
#   2004/12/11 19:23:14+01:00 spock@spock.one.pl 
#   Fixed color corruption problem in DIRECTCOLOR modes.
# 
# drivers/video/console/fbcon.c
#   2004/12/11 19:23:01+01:00 spock@spock.one.pl +4 -3
#   Fixed color corruption problem in DIRECTCOLOR modes.
# 
# ChangeSet
#   2004/12/11 19:18:14+01:00 spock@spock.one.pl 
#   Fixed a problem with text scrolling in silent mode.
# 
# drivers/video/console/fbcon.c
#   2004/12/11 19:18:03+01:00 spock@spock.one.pl +1 -1
#   Fixed a problem with text scrolling in silent mode.
# 
# ChangeSet
#   2004/11/22 17:51:30+01:00 spock@spock.one.pl 
#   Merged upstream changes.
# 
# drivers/video/console/fbcon.c
#   2004/11/22 17:51:20+01:00 spock@spock.one.pl +2 -1
#   Merged upstream changes.
# 
# drivers/video/cfbsplash.c
#   2004/11/22 17:51:20+01:00 spock@spock.one.pl +18 -38
#   Merged upstream changes.
# 
# ChangeSet
#   2004/11/21 13:08:08+01:00 spock@spock.one.pl 
#   Merged upstream changes.
# 
# kernel/sysctl.c
#   2004/11/21 13:07:57+01:00 spock@spock.one.pl +2 -2
#   Merged upstream changes.
# 
# kernel/panic.c
#   2004/11/21 13:07:57+01:00 spock@spock.one.pl +0 -0
#   Merged upstream changes.
# 
# drivers/video/console/fbcon.c
#   2004/11/21 13:07:57+01:00 spock@spock.one.pl +25 -16
#   Merged upstream changes.
# 
# drivers/video/console/bitblit.c
#   2004/11/21 13:07:57+01:00 spock@spock.one.pl +2 -2
#   Merged upstream changes.
# 
# include/linux/fb.h
#   2004/11/21 12:41:35+01:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# ChangeSet
#   2004/10/30 22:07:03+02:00 spock@spock.one.pl 
#   Merged upstream changes.
# 
# drivers/video/console/fbcon.c
#   2004/10/30 22:06:53+02:00 spock@spock.one.pl +0 -1
#   Merged upstream changes.
# 
# drivers/video/console/bitblit.c
#   2004/10/30 22:06:53+02:00 spock@spock.one.pl +5 -11
#   Merged upstream changes.
# 
# kernel/sysctl.c
#   2004/10/30 21:26:34+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# include/linux/fb.h
#   2004/10/30 21:26:34+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# drivers/video/Makefile
#   2004/10/30 21:26:34+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# drivers/video/Kconfig
#   2004/10/30 21:26:34+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# drivers/Makefile
#   2004/10/30 21:26:34+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# ChangeSet
#   2004/10/28 19:24:43+02:00 spock@spock.one.pl 
#   Merged updated fbsplash code.
# 
# drivers/video/console/fbcon.c
#   2004/10/28 19:24:35+02:00 spock@spock.one.pl +129 -9
#   Merged updated fbsplash code.
# 
# drivers/video/console/bitblit.c
#   2004/10/28 19:24:35+02:00 spock@spock.one.pl +18 -2
#   Moved some fbsplash code to bitblit.c.
# 
# drivers/video/Kconfig
#   2004/10/28 19:24:35+02:00 spock@spock.one.pl +1 -1
#   Added a dependency on standard bit-blitting.
# 
# ChangeSet
#   2004/10/28 19:15:14+02:00 spock@spock.one.pl 
#   Merged upstream changes.
# 
# include/linux/fb.h
#   2004/10/28 19:15:05+02:00 spock@spock.one.pl +2 -1
#   Merged upstream changes.
# 
# ChangeSet
#   2004/10/28 19:10:45+02:00 spock@spock.one.pl 
#   Merged upstream changes.
# 
# drivers/video/console/fbcon.c
#   2004/10/28 19:10:37+02:00 spock@spock.one.pl +6 -119
#   Resolved some conflicts.
# 
# kernel/sysctl.c
#   2004/10/28 00:41:24+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# kernel/power/console.c
#   2004/10/28 00:41:24+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# kernel/panic.c
#   2004/10/28 00:41:24+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# include/linux/sysctl.h
#   2004/10/28 00:41:24+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# drivers/video/Kconfig
#   2004/10/28 00:41:24+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# drivers/char/n_tty.c
#   2004/10/28 00:41:24+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# drivers/Makefile
#   2004/10/28 00:41:24+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# ChangeSet
#   2004/10/21 22:21:56+02:00 spock@spock.one.pl 
#   Fixed a little big-endian bug.
# 
# drivers/video/cfbsplash.c
#   2004/10/21 22:21:42+02:00 spock@spock.one.pl +1 -1
#   Fixed a little big-endian bug. 
# 
# ChangeSet
#   2004/09/30 16:42:39+02:00 spock@spock.one.pl 
#   Merged conflicts.
# 
# drivers/char/n_tty.c
#   2004/09/30 16:42:27+02:00 spock@spock.one.pl +9 -0
#   Merged conflicts.
# 
# drivers/Makefile
#   2004/09/30 16:42:27+02:00 spock@spock.one.pl +2 -3
#   Merged conflicts.
# 
# include/linux/sysctl.h
#   2004/09/30 16:35:14+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# include/linux/fb.h
#   2004/09/30 16:35:14+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# drivers/video/console/fbcon.c
#   2004/09/30 16:35:14+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# drivers/video/Makefile
#   2004/09/30 16:35:13+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# drivers/char/keyboard.c
#   2004/09/30 16:35:13+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# ChangeSet
#   2004/09/26 11:16:27+02:00 spock@spock.one.pl 
#   Made get_color visible outside fbcon.c
# 
# drivers/video/console/fbcon.c
#   2004/09/26 11:16:18+02:00 spock@spock.one.pl +2 -2
#   Made get_color visible outside fbcon.c
# 
# ChangeSet
#   2004/09/26 03:21:27+02:00 spock@spock.one.pl 
#   Updated fbsplash to work with 2.6.9-rc2, possibly fixing the random chars problem along the way.
# 
# drivers/video/fbsplash.c
#   2004/09/26 03:21:01+02:00 spock@spock.one.pl +2 -0
#   Updated to work with 2.6.9-rc2.
# 
# drivers/video/console/fbcon.c
#   2004/09/26 03:21:01+02:00 spock@spock.one.pl +2 -2
#   Updated to work with 2.6.9-rc2.
# 
# drivers/video/cfbsplash.c
#   2004/09/26 03:21:01+02:00 spock@spock.one.pl +34 -26
#   Changed cursor render code, this should correct the problems with strange characters filling empty spaces on the screen some people were experiencing.
# 
# drivers/Makefile
#   2004/09/26 03:21:01+02:00 spock@spock.one.pl +1 -1
#   Changed the order of loading drivers - we need tty0 to be registered when fbsplash_init is called.
# 
# ChangeSet
#   2004/09/14 20:12:46+02:00 spock@spock.one.pl 
#   Merged fbsplash conflicts.
# 
# drivers/video/console/fbcon.c
#   2004/09/14 20:12:37+02:00 spock@spock.one.pl +0 -1
#   Merged fbsplash conflicts.
# 
# include/linux/fb.h
#   2004/09/14 19:51:33+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# drivers/video/Makefile
#   2004/09/14 19:51:33+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# ChangeSet
#   2004/09/07 15:12:03+02:00 spock@spock.one.pl 
#   Fixed the code to handle modes in which line_length != xres * bytespp.
#   TAG: fbsplash-0.9-r8
# 
# drivers/video/fbsplash.h
#   2004/09/07 15:11:52+02:00 spock@spock.one.pl +1 -1
#   Fixed the code to handle modes in which line_length != xres * bytespp.
# 
# drivers/video/cfbsplash.c
#   2004/09/07 15:11:52+02:00 spock@spock.one.pl +13 -10
#   Fixed the code to handle modes in which line_length != xres * bytespp.
# 
# ChangeSet
#   2004/09/07 15:05:13+02:00 spock@spock.one.pl 
#   Merge spock.one.pl:/devel/kernel/linux-2.5
#   into spock.one.pl:/devel/kernel/linux-fbsplash
# 
# kernel/sysctl.c
#   2004/09/07 15:05:07+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# include/linux/sysctl.h
#   2004/09/07 15:05:06+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# drivers/video/Makefile
#   2004/09/07 15:05:06+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# ChangeSet
#   2004/08/26 21:33:29+02:00 spock@spock.one.pl 
#   Added a fix to prevent a deadlock when using swsusp resume and fbsplash in silent mode.
#   TAG: fbsplash-0.9-r7
# 
# drivers/video/fbsplash.h
#   2004/08/26 21:30:50+02:00 spock@spock.one.pl +7 -1
#   Added the fbsplash_issilent() macro.
# 
# kernel/power/console.c
#   2004/08/26 21:29:07+02:00 spock@spock.one.pl +5 -0
#   Added a fix to prevent a deadlock when using swsusp resume and fbsplash in silent mode.
# 
# drivers/video/console/fbcon.c
#   2004/08/24 19:01:18+02:00 spock@spock.one.pl +6 -50
#   It seems I've screwed something up while merging the fbcon code. It's fixed now.
# 
# ChangeSet
#   2004/08/24 18:55:58+02:00 spock@spock.one.pl 
#   Resolved conflicts with fbcon and sysctl.
# 
# include/linux/sysctl.h
#   2004/08/24 18:55:47+02:00 spock@spock.one.pl +1 -1
#   Resolved conflicts.
# 
# drivers/video/console/fbcon.c
#   2004/08/24 18:55:47+02:00 spock@spock.one.pl +49 -5
#   Resolved conflicts.
# 
# kernel/sysctl.c
#   2004/08/24 18:38:09+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# kernel/panic.c
#   2004/08/24 18:38:08+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# include/linux/fb.h
#   2004/08/24 18:38:08+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# drivers/video/Makefile
#   2004/08/24 18:38:08+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# drivers/video/Kconfig
#   2004/08/24 18:38:08+02:00 spock@spock.one.pl +0 -0
#   Auto merged
# 
# ChangeSet
#   2004/08/23 11:55:52+02:00 spock@spock.one.pl 
#   Changed the order of acquiring console_sem - the sem has to be held before doing fbsplash_disable to avoid a lot of ugly badnesses.
#   TAG: fbsplash-0.9-r6
# 
# drivers/video/fbsplash.c
#   2004/08/21 20:49:21+02:00 spock@spock.one.pl +1 -1
#   Changed the order of acquiring console_sem - the sem has to be held before doing fbsplash_disable
# 
# BitKeeper/etc/config
#   2004/08/19 00:22:09+02:00 spock@spock.one.pl +2 -3
#   changed configuration settings for fbsplash
# 
# BitKeeper/etc/logging_ok
#   2004/08/19 00:21:58+02:00 spock@spock.one.pl +0 -6
#   changed configuration settings for fbsplash
# 
# ChangeSet
#   2004/08/18 23:57:21+02:00 spock@spock.one.pl 
#   Imported fbsplash-0.9-r5
# 
# kernel/panic.c
#   2004/08/18 23:57:12+02:00 spock@spock.one.pl +5 -0
#   Imported fbsplash-0.9-r5
# 
# include/linux/sysctl.h
#   2004/08/18 23:57:12+02:00 spock@spock.one.pl +1 -0
#   Imported fbsplash-0.9-r5
# 
# include/linux/fb.h
#   2004/08/18 23:57:12+02:00 spock@spock.one.pl +22 -1
#   Imported fbsplash-0.9-r5
# 
# include/linux/console_struct.h
#   2004/08/18 23:57:12+02:00 spock@spock.one.pl +3 -0
#   Imported fbsplash-0.9-r5
# 
# drivers/video/console/fbcon.c
#   2004/08/18 23:57:12+02:00 spock@spock.one.pl +160 -10
#   Imported fbsplash-0.9-r5
# 
# drivers/video/Makefile
#   2004/08/18 23:57:12+02:00 spock@spock.one.pl +1 -0
#   Imported fbsplash-0.9-r5
# 
# drivers/video/Kconfig
#   2004/08/18 23:57:12+02:00 spock@spock.one.pl +11 -1
#   Imported fbsplash-0.9-r5
# 
# drivers/char/n_tty.c
#   2004/08/18 23:57:12+02:00 spock@spock.one.pl +11 -0
#   Imported fbsplash-0.9-r5
# 
# drivers/char/keyboard.c
#   2004/08/18 23:57:12+02:00 spock@spock.one.pl +8 -0
#   Imported fbsplash-0.9-r5
# 
# Documentation/fb/00-INDEX
#   2004/08/18 23:57:12+02:00 spock@spock.one.pl +2 -0
#   Imported fbsplash-0.9-r5
# 
# ChangeSet
#   2004/08/18 23:54:55+02:00 spock@spock.one.pl 
#   Imported fbsplash-0.9-r5
# 
# kernel/sysctl.c
#   2004/08/18 23:54:46+02:00 spock@spock.one.pl +14 -0
#   Imported fbsplash-0.9-r5
# 
# include/linux/console_splash.h
#   2004/08/18 23:50:32+02:00 spock@spock.one.pl +13 -0
# 
# drivers/video/fbsplash.h
#   2004/08/18 23:50:32+02:00 spock@spock.one.pl +69 -0
# 
# drivers/video/fbsplash.c
#   2004/08/18 23:50:32+02:00 spock@spock.one.pl +527 -0
# 
# drivers/video/cfbsplash.c
#   2004/08/18 23:50:32+02:00 spock@spock.one.pl +471 -0
# 
# include/linux/console_splash.h
#   2004/08/18 23:50:32+02:00 spock@spock.one.pl +0 -0
#   BitKeeper file /devel/kernel/linux-fbsplash/include/linux/console_splash.h
# 
# drivers/video/fbsplash.h
#   2004/08/18 23:50:32+02:00 spock@spock.one.pl +0 -0
#   BitKeeper file /devel/kernel/linux-fbsplash/drivers/video/fbsplash.h
# 
# drivers/video/fbsplash.c
#   2004/08/18 23:50:32+02:00 spock@spock.one.pl +0 -0
#   BitKeeper file /devel/kernel/linux-fbsplash/drivers/video/fbsplash.c
# 
# drivers/video/cfbsplash.c
#   2004/08/18 23:50:32+02:00 spock@spock.one.pl +0 -0
#   BitKeeper file /devel/kernel/linux-fbsplash/drivers/video/cfbsplash.c
# 
# Documentation/fb/splash.txt
#   2004/08/18 23:50:31+02:00 spock@spock.one.pl +230 -0
# 
# Documentation/fb/splash.txt
#   2004/08/18 23:50:31+02:00 spock@spock.one.pl +0 -0
#   BitKeeper file /devel/kernel/linux-fbsplash/Documentation/fb/splash.txt
# 
diff -Nru a/Documentation/fb/00-INDEX b/Documentation/fb/00-INDEX
--- a/Documentation/fb/00-INDEX	2005-02-13 14:35:45 +01:00
+++ b/Documentation/fb/00-INDEX	2005-02-13 14:35:45 +01:00
@@ -19,6 +19,8 @@
 	- info on the Matrox frame buffer driver
 pvr2fb.txt
 	- info on the PowerVR 2 frame buffer driver
+splash.txt
+	- info on the Framebuffer Splash
 tgafb.txt
 	- info on the TGA (DECChip 21030) frame buffer driver
 vesafb.txt
diff -Nru a/Documentation/fb/splash.txt b/Documentation/fb/splash.txt
--- /dev/null	Wed Dec 31 16:00:00 196900
+++ b/Documentation/fb/splash.txt	2005-02-13 14:35:45 +01:00
@@ -0,0 +1,230 @@
+What is it?
+-----------
+
+The framebuffer splash is a kernel feature that allows displaying a background
+picture on selected consoles and switching the first console to the so-called
+silent mode, while booting/rebooting/shutting down the system.
+
+What do I need to get it to work?
+---------------------------------
+
+To get fb splash up-and-running you will have to:
+ 1) get a copy of splashutils from (FIXME URL)
+ 2) get some splash themes (FIXME URL)
+ 3) build the kernel helper program
+ 4) build your kernel with the FB_SPLASH option enabled.
+
+To get fb splash operational right after fbcon initialization is finished, you
+will have to include a theme and the kernel helper into your initramfs image.
+Please refer to splashutils documentation for instructions on how to do that.
+
+Operation modes
+---------------
+
+The framebuffer splash can work in two modes: verbose and silent. The first
+one means "console with a background image". The latter one is a concept
+first introduced in bootsplash. When silent mode is active, the console is
+switched to a graphic mode and no text is displayed. It's up to the userspace
+programs to display something (a progress bar for example) on the screen. 
+
+Kernel command line parameters
+------------------------------
+
+Framebuffer splash can be configured from the kernel command line by passing
+options in the following way: "splash=option1,option2".
+
+The following options are recognized:
+
+off     - do not enable fbsplash after fbcon initialization; this is
+          default behaviour
+verbose - switch fbsplash to verbose mode after fbcon is initialized
+silent  - switch fbsplash to silent mode after fbcon is initialized
+
+theme:<name> - use theme 'name' on the first console, default theme name
+               is 'default' (who'd have guessed?)
+
+Example - you want to get verbose splash with the theme 'tux' right after 
+fbcon is up:
+  splash=verbose,theme:tux
+
+The userspace helper
+--------------------
+
+The userspace splash helper (by default: /sbin/splash_helper) is called by the
+kernel whenever an important event occurs and the kernel needs some kind of
+job to be carried out. Important events include console switches and graphic
+mode switches (the kernel requests background images and config for the current 
+console). The splash helper must be accessible at all times. If it's not,
+fbsplash will be switched off automatically.
+
+It's possible to set path to the splash helper by writing it to 
+/proc/sys/kernel/fbsplash.
+
+*****************************************************************************
+
+The information below is mostly technical stuff. If you don't plan to develop
+something fbsplash-related, there's probably no need to read it.
+
+The splash protocol
+-------------------
+
+The splash protocol defines a communication interface between the kernel and 
+the userspace splash helper.
+
+The kernel side is responsible for:
+
+ o rendering console text, using an image as a background (instead of a
+   standard solid color fbcon uses),
+ o accepting commands from the user via ioctls on the fbsplash device,
+ o calling the userspace helper to set things up as soon as the fb subsystem is
+   initialized.
+
+The userspace helper is responsible for everything else, including parsing
+configuration files, decompressing the image files whenever the kernel needs
+it, and communicating with the kernel if necessary.
+
+The splash protocol specifies how communication is done in both ways:
+kernel->userspace and userspace->helper.
+  
+Kernel -> Userspace
+-------------------
+
+The kernel communicates with the userspace helper by calling it and specifying
+the task that to be done in a series of arguments.
+
+The arguments follow the pattern:
+<splash protocol version> <command> <parameters>
+
+All commands defined in splash protocol v1 have the following parameters:
+ virtual console
+ framebuffer number
+ splash mode - 'v' indicates verbose, 's' indicates silent
+ theme
+
+Splash protocol v1 specifies the following commands:
+
+getpic
+------
+ The kernel issues this command to request image data. It's up to the userspace
+ helper to find a background image appropriate for the specified theme and the 
+ current resolution. The userspace helper should respond by issuing the
+ FBIOSPLASH_SETPIC ioctl.
+
+init
+----
+ The kernel issues this command if either 'verbose' or 'silent' has been
+ specified in splash= on the kernel command line. Upon receiving this command,
+ the userspace helper should find a proper configuration file for the
+ specified theme and the current resolution, and issue the FBIOSPLASH_SETCFG.
+ FBIOSPLASH_SETPIC and FBIOSPLASH_SETSTATE commands.
+
+ When the userspace helper is called in an early phase of the boot process
+ (right after the initialization of fbcon), no filesystems will be mounted.
+ The helper program should mount sysfs and then create the appropriate 
+ framebuffer, fbsplash and tty0 devices (if they don't already exist) to get 
+ current display settings and to be able to communicate with the kernel side.
+
+ Note that the console sem is not held when the kernel calls splash_helper
+ with the 'init' command. The splash helper should perform all ioctls with
+ origin set to FB_SPLASH_IO_ORIG_USER.
+
+modechange
+----------
+ The kernel issues this command on a mode change. The helper's response should
+ be similar to the response to the 'init' command. Note that this time the
+ console sem is held and all ioctls must be performed with origin set to
+ FB_SPLASH_IO_ORIG_KERNEL.
+
+
+Userspace -> Kernel
+-------------------
+
+Userspace programs can communicate with the kernel side via ioctls on the 
+on the fbsplash device. These ioctls are to be used by both the userspace helper 
+(called only by the kernel) and userspace configuration tools (run by the users).
+The splash helper should set the origin field to FB_SPLASH_IO_ORIG_KERNEL when
+doing the appropriate ioctls. All userspace configuration tools should use
+FB_SPLASH_IO_ORIG_USER. Failure to set the appropriate value in the origin
+field when performing ioctls from the kernel helper may result in a deadlock.
+
+The framebuffer splash provides the following ioctls (all defined in linux/fb.h):
+
+FBIOSPLASH_SETMODE
+description: sets the global fbsplash mode
+argument: unsigned int; values: FB_SPLASH_MODE_VERBOSE, FB_SPLASH_MODE_SILENT
+
+FBIOSPLASH_GETMODE
+description: gets the global fbsplash mode 
+argument: unsigned int*; values as in FBIOSPLASH_SETMODE
+
+FBIOSPLASH_SETPIC
+description: loads a background picture for a virtual console
+argument: struct fb_splash_iowrapper*; data: struct fb_image*
+notes: 
+If called for consoles other than the current foreground one, the picture data 
+will be ignored.
+
+If the current virtual console is running in a 8-bpp mode, the cmap substruct
+of fb_image has to be filled appropriately: start should be set to 16 (first 16 
+colors are reserved for fbcon), len to a value <= 240 and red, green and blue
+should point to valid cmap data. The transp field is ingored. The fields dx, dy
+bg_color, fg_color in fb_image are ignored as well.
+
+FBIOSPLASH_SETCFG
+description: sets the fbsplash config for a virtual console
+argument: struct fb_splash_iowrapper*; data: struct vc_splash*; 
+          the structure has to be filled with valid data.
+
+FBIOSPLASH_GETCFG
+description: gets the fbsplash config for a virtual console
+argument: struct fb_splash_iowrapper*; data: struct vc_splash*
+
+FBIOSPLASH_SETSTATE
+description: sets the fbsplash state for a virtual console
+argument: struct fb_splash_iowrapper*; data: unsigned int*
+	  values: 0 = disabled, 1 = enabled.
+
+FBIOSPLASH_GETSTATE
+description: gets the fbsplash state for a virtual console
+argument: struct fb_splash_iowrapper*; data: unsigned int*
+          values: as in FBIOSPLASH_SETSTATE
+
+Info on used structures:
+
+Definition of struct vc_splash can be found in linux/console_splash.h, it's
+heavily commented, so no special descriptions are necessary. Note that 
+the 'theme' field should point to a string no longer than FB_SPLASH_THEME_LEN.
+When FBIOSPLASH_GETCFG call is performed, the theme field should point to a
+char buffer of length FB_SPLASH_THEME_LEN.
+
+Definition of struct fb_splash_iowrapper can be found in linux/fb.h. 
+The fields in this struct have the following meaning:
+
+vc: 
+Virtual console number.
+
+origin: 
+Specifies if the ioctl is performed as a response to a kernel request. The
+splash helper should set this field to FB_SPLASH_IO_ORIG_KERNEL, userspace
+programs should set it to FB_SPLASH_IO_ORIG_USER. This field is necessary to
+avoid console semaphore deadlocks.
+
+data: 
+Pointer to a data structure appropriate for the performed ioctl. Type of
+the data struct is specified in the ioctls description, for example 'data:
+struct vc_splash*' means that the 'data' field should be a pointer to a
+vc_splash struct.
+
+*****************************************************************************
+
+Credit
+------
+
+Original idea & implementation by:
+  Volker Poplawski <volker@poplawski.de>, Stefan Reinauer <stepan@suse.de>,
+  Steffen Winterfeldt <snwint@suse.de>, Michael Schroeder <mls@suse.de>,
+  Ken Wimer <wimer@suse.de>.
+
+Splash protocol redesign, current implementation, docs by:
+  Michal Januszewski <spock@gentoo.org>
+
diff -Nru a/drivers/Makefile b/drivers/Makefile
--- a/drivers/Makefile	2005-02-13 14:35:45 +01:00
+++ b/drivers/Makefile	2005-02-13 14:35:45 +01:00
@@ -7,15 +7,14 @@
 
 obj-$(CONFIG_PCI)		+= pci/
 obj-$(CONFIG_PARISC)		+= parisc/
+# char/ comes before serial/ etc so that the VT console is the boot-time
+# default.
+obj-y				+= char/
 obj-y				+= video/
 obj-$(CONFIG_ACPI_BOOT)		+= acpi/
 # PnP must come after ACPI since it will eventually need to check if acpi
 # was used and do nothing if so
 obj-$(CONFIG_PNP)		+= pnp/
-
-# char/ comes before serial/ etc so that the VT console is the boot-time
-# default.
-obj-y				+= char/
 
 # i810fb and intelfb depend on char/agp/
 obj-$(CONFIG_FB_I810)           += video/i810/
diff -Nru a/drivers/char/keyboard.c b/drivers/char/keyboard.c
--- a/drivers/char/keyboard.c	2005-02-13 14:35:45 +01:00
+++ b/drivers/char/keyboard.c	2005-02-13 14:35:45 +01:00
@@ -40,6 +40,8 @@
 #include <linux/sysrq.h>
 #include <linux/input.h>
 
+#include "../video/fbsplash.h"
+
 static void kbd_disconnect(struct input_handle *handle);
 extern void ctrl_alt_del(void);
 
@@ -1057,6 +1059,12 @@
 		if (emulate_raw(vc, keycode, !down << 7))
 			if (keycode < BTN_MISC)
 				printk(KERN_WARNING "keyboard.c: can't emulate rawmode for keycode %d\n", keycode);
+
+	/* switch splash to verbose mode if ESC or F2 is pressed */
+	if (down == 1 && (keycode == KEY_ESC || keycode == KEY_F2)) {
+		if (fbsplash_verbose())
+			return;
+	}
 
 #ifdef CONFIG_MAGIC_SYSRQ	       /* Handle the SysRq Hack */
 	if (keycode == KEY_SYSRQ && (sysrq_down || (down == 1 && sysrq_alt))) {
diff -Nru a/drivers/char/n_tty.c b/drivers/char/n_tty.c
--- a/drivers/char/n_tty.c	2005-02-13 14:35:45 +01:00
+++ b/drivers/char/n_tty.c	2005-02-13 14:35:45 +01:00
@@ -49,6 +49,8 @@
 #include <asm/uaccess.h>
 #include <asm/system.h>
 
+#include "../video/fbsplash.h"
+
 /* number of characters left in xmit buffer before select has we have room */
 #define WAKEUP_CHARS 256
 
@@ -1179,6 +1181,15 @@
  
 static int job_control(struct tty_struct *tty, struct file *file)
 {
+	/* automatically switch splash to verbose mode if someone tries to 
+	   read from tty0 */
+	if (file->f_dentry->d_inode->i_rdev == MKDEV(TTY_MAJOR,0) ||
+	    file->f_dentry->d_inode->i_rdev == MKDEV(TTY_MAJOR,1) ||
+	    file->f_dentry->d_inode->i_rdev == MKDEV(TTYAUX_MAJOR,0) ||	
+	    file->f_dentry->d_inode->i_rdev == MKDEV(TTYAUX_MAJOR,1)) {
+		fbsplash_verbose();
+	}
+
 	/* Job control check -- must be done at start and after
 	   every sleep (POSIX.1 7.1.1.4). */
 	/* NOTE: not yet done after every sleep pending a thorough
diff -Nru a/drivers/video/Kconfig b/drivers/video/Kconfig
--- a/drivers/video/Kconfig	2005-02-13 14:35:45 +01:00
+++ b/drivers/video/Kconfig	2005-02-13 14:35:45 +01:00
@@ -1140,5 +1140,15 @@
 	source "drivers/video/backlight/Kconfig"
 endif
 
-endmenu
+config FB_SPLASH
+	bool "Support for the framebuffer splash"
+	depends on FRAMEBUFFER_CONSOLE=y && !FB_TILEBLITTING
+	default n
+	---help---
+	  This option enables support for the Linux boot-up splash screen and
+	  graphical backgrounds on consoles. Note that you will need userspace
+	  splashutils in order to take advantage of these features. Refer to
+	  Documentation/fb/splash.txt for more information.
 
+	  If unsure, say N.
+endmenu
diff -Nru a/drivers/video/Makefile b/drivers/video/Makefile
--- a/drivers/video/Makefile	2005-02-13 14:35:45 +01:00
+++ b/drivers/video/Makefile	2005-02-13 14:35:45 +01:00
@@ -7,6 +7,7 @@
 obj-$(CONFIG_VT)		  += console/
 obj-$(CONFIG_LOGO)		  += logo/
 obj-$(CONFIG_SYSFS)		  += backlight/
+obj-$(CONFIG_FB_SPLASH)           += fbsplash.o cfbsplash.o
 
 obj-$(CONFIG_FB)                  += fbmem.o fbmon.o fbcmap.o fbsysfs.o modedb.o softcursor.o
 # Only include macmodes.o if we have FB support and are PPC
diff -Nru a/drivers/video/cfbsplash.c b/drivers/video/cfbsplash.c
--- /dev/null	Wed Dec 31 16:00:00 196900
+++ b/drivers/video/cfbsplash.c	2005-02-13 14:35:45 +01:00
@@ -0,0 +1,468 @@
+/*
+ *  linux/drivers/video/cfbsplash.c -- Framebuffer splash render functions
+ *  
+ *  Copyright (C) 2004 Michal Januszewski <spock@gentoo.org>
+ *
+ *  Code based upon "Bootsplash" (C) 2001-2003 
+ *       Volker Poplawski <volker@poplawski.de>,
+ *       Stefan Reinauer <stepan@suse.de>,
+ *       Steffen Winterfeldt <snwint@suse.de>,
+ *       Michael Schroeder <mls@suse.de>,
+ *       Ken Wimer <wimer@suse.de>.
+ *
+ *  This file is subject to the terms and conditions of the GNU General Public
+ *  License.  See the file COPYING in the main directory of this archive for
+ *  more details.
+ */ 
+#include <linux/config.h>
+#include <linux/module.h>
+#include <linux/types.h>
+#include <linux/fb.h>
+#include <linux/selection.h>
+#include <linux/vt_kern.h>
+#include <asm/irq.h>
+#include <asm/system.h>
+
+#include "console/fbcon.h"
+#include "fbsplash.h"
+
+#define parse_pixel(shift,bpp,type)						\
+	do {									\
+		if (d & (0x80 >> (shift)))					\
+			dd2[(shift)] = fgx;					\
+		else								\
+			dd2[(shift)] = transparent ? *(type *)splash_src : bgx;	\
+		splash_src += (bpp);						\
+	} while (0)								\
+
+extern int get_color(struct vc_data *vc, struct fb_info *info,
+		     u16 c, int is_fg);
+
+void fbsplash_fix_pseudo_pal(struct fb_info *info, struct vc_data *vc)
+{
+	int i, j, k;
+	int minlen = min(min(info->var.red.length, info->var.green.length), 
+			     info->var.blue.length);
+	u32 col;
+	
+	for (j = i = 0; i < 16; i++) {
+		k = color_table[i];
+                      
+		col = ((vc->vc_palette[j++]  >> (8-minlen)) 
+			<< info->var.red.offset);
+		col |= ((vc->vc_palette[j++] >> (8-minlen)) 
+			<< info->var.green.offset);
+		col |= ((vc->vc_palette[j++] >> (8-minlen)) 
+			<< info->var.blue.offset);
+			((u32 *)info->pseudo_palette)[k] = col;
+	}
+}
+				
+void fbsplash_renderc(struct fb_info *info, int ypos, int xpos, int height, 
+		      int width, u8* src, u32 fgx, u32 bgx, u8 transparent)
+{	
+	unsigned int x, y;
+	u32 dd;
+	int bytespp = ((info->var.bits_per_pixel + 7) >> 3);
+	unsigned int d = ypos * info->fix.line_length + xpos * bytespp;
+	unsigned int ds = (ypos * info->var.xres + xpos) * bytespp;
+	u16 dd2[4];
+
+	u8* splash_src = (u8 *)(info->splash.data + ds);
+	u8* dst = (u8 *)(info->screen_base + d);
+
+	if ((ypos + height) > info->var.yres || (xpos + width) > info->var.xres)
+		return;
+	
+	for (y = 0; y < height; y++) {
+		switch (info->var.bits_per_pixel) {
+	
+		case 32:
+			for (x = 0; x < width; x++) {
+
+				if ((x & 7) == 0)
+					d = *src++;
+				if (d & 0x80)
+					dd = fgx;
+				else
+					dd = transparent ? 
+					     *(u32 *)splash_src : bgx;
+				
+				d <<= 1;
+				splash_src += 4;
+				fb_writel(dd, dst);
+				dst += 4;
+			}
+			break;
+		case 24:
+			for (x = 0; x < width; x++) {
+
+				if ((x & 7) == 0)
+					d = *src++;
+				if (d & 0x80)
+					dd = fgx;
+				else
+					dd = transparent ? 
+					     (*(u32 *)splash_src & 0xffffff) : bgx;
+				
+				d <<= 1;
+				splash_src += 3;
+#ifdef __LITTLE_ENDIAN
+				fb_writew(dd & 0xffff, dst);
+				dst += 2;
+				fb_writeb((dd >> 16), dst);
+#else
+				fb_writew(dd >> 8, dst);
+				dst += 2;
+				fb_writeb(dd & 0xff, dst);
+#endif
+				dst++;
+			}
+			break;
+		case 16:
+			for (x = 0; x < width; x += 2) {
+		    		if ((x & 7) == 0)
+					d = *src++;
+
+				parse_pixel(0, 2, u16);
+				parse_pixel(1, 2, u16);
+#ifdef __LITTLE_ENDIAN
+				dd = dd2[0] | (dd2[1] << 16);
+#else
+				dd = dd2[1] | (dd2[0] << 16);
+#endif
+				d <<= 2;
+				fb_writel(dd, dst);
+				dst += 4;
+			}
+			break;
+
+		case 8:
+			for (x = 0; x < width; x += 4) {
+				if ((x & 7) == 0)
+					d = *src++;
+	
+				parse_pixel(0, 1, u8);
+				parse_pixel(1, 1, u8);
+				parse_pixel(2, 1, u8);
+				parse_pixel(3, 1, u8);
+		
+#ifdef __LITTLE_ENDIAN
+				dd = dd2[0] | (dd2[1] << 8) | (dd2[2] << 16) | (dd2[3] << 24);
+#else
+				dd = dd2[3] | (dd2[2] << 8) | (dd2[1] << 16) | (dd2[0] << 24);
+#endif
+				d <<= 4;
+				fb_writel(dd, dst);
+				dst += 4;
+			}		
+		}
+
+		dst += info->fix.line_length - width * bytespp;
+		splash_src += (info->var.xres - width) * bytespp;
+    	}
+}
+
+#define cc2cx(a) 						\
+	((info->fix.visual == FB_VISUAL_TRUECOLOR || 		\
+	  info->fix.visual == FB_VISUAL_DIRECTCOLOR) ? 		\
+	 ((u32*)info->pseudo_palette)[a] : a)
+
+void fbsplash_putcs(struct vc_data *vc, struct fb_info *info,
+		   const unsigned short *s, int count, int yy, int xx)
+{
+	unsigned short charmask = vc->vc_hi_font_mask ? 0x1ff : 0xff;
+	int fg_color, bg_color, transparent;
+
+	u8 *src;
+	u32 bgx, fgx;
+
+	u16 c = scr_readw(s);
+
+	fg_color = get_color(vc, info, c, 1);
+        bg_color = get_color(vc, info, c, 0);
+	transparent = vc->vc_splash.bg_color == bg_color;
+
+	xx = xx * vc->vc_font.width + vc->vc_splash.tx;
+	yy = yy * vc->vc_font.height + vc->vc_splash.ty;
+
+	fgx = cc2cx(fg_color);
+	bgx = cc2cx(bg_color);
+
+	while (count--) {
+		c = scr_readw(s++);
+		src = vc->vc_font.data + (c & charmask) * vc->vc_font.height *
+		      ((vc->vc_font.width + 7) >> 3);
+
+		fbsplash_renderc(info, yy, xx, vc->vc_font.height, 
+			       vc->vc_font.width, src, fgx, bgx, transparent);
+		xx += vc->vc_font.width;
+	}
+}
+
+void fbsplash_cursor(struct fb_info *info, struct fb_cursor *cursor)
+{
+	int i;
+	unsigned int dsize, s_pitch;
+	struct fbcon_ops *ops = info->fbcon_par;
+	struct vc_data* vc;	
+	u8 *src;
+
+	/* we really don't need any cursors while the console is blanked */
+	if (info->state != FBINFO_STATE_RUNNING || ops->blank_state)
+		return;
+
+	vc = vc_cons[ops->currcon].d;
+
+	src = kmalloc(64 + sizeof(struct fb_image), GFP_ATOMIC);
+	if (!src)
+		return;
+
+	s_pitch = (cursor->image.width + 7) >> 3;
+	dsize = s_pitch * cursor->image.height;
+	if (cursor->enable) {	
+		switch (cursor->rop) {
+		case ROP_XOR:
+			for (i = 0; i < dsize; i++)
+				src[i] = cursor->image.data[i] ^ cursor->mask[i];
+                        break;
+		case ROP_COPY:
+		default:
+			for (i = 0; i < dsize; i++)
+				src[i] = cursor->image.data[i] & cursor->mask[i];
+			break;
+		}
+	} else
+		memcpy(src, cursor->image.data, dsize);
+
+	fbsplash_renderc(info,
+			cursor->image.dy + vc->vc_splash.ty,
+			cursor->image.dx + vc->vc_splash.tx,
+			cursor->image.height,
+			cursor->image.width,
+			(u8*)src,
+			cc2cx(cursor->image.fg_color),
+			cc2cx(cursor->image.bg_color),
+			cursor->image.bg_color == vc->vc_splash.bg_color);
+
+	kfree(src);
+}
+
+static void splashset(u8 *dst, int height, int width, int dstbytes, 
+		        u32 bgx, int bpp)
+{
+	int i;
+	
+	if (bpp == 8)
+		bgx |= bgx << 8;
+	if (bpp == 16 || bpp == 8)
+		bgx |= bgx << 16;
+	
+	while (height-- > 0) {
+		u8 *p = dst;
+		
+		switch (bpp) {
+		
+		case 32:
+			for (i=0; i < width; i++) {
+				fb_writel(bgx, p); p += 4;
+			}
+			break;
+		case 24:	
+			for (i=0; i < width; i++) {
+#ifdef __LITTLE_ENDIAN
+				fb_writew((bgx & 0xffff),(u16*)p); p += 2;
+				fb_writeb((bgx >> 16),p++);
+#else
+				fb_writew((bgx >> 8),(u16*)p); p += 2;
+				fb_writeb((bgx & 0xff),p++);
+#endif
+			}
+		case 16:
+			for (i=0; i < width/4; i++) {
+				fb_writel(bgx,p); p += 4;
+				fb_writel(bgx,p); p += 4;
+			}
+			if (width & 2) {
+				fb_writel(bgx,p); p += 4;
+			}
+			if (width & 1)
+				fb_writew(bgx,(u16*)p);
+			break;
+		case 8:
+			for (i=0; i < width/4; i++) {
+				fb_writel(bgx,p); p += 4;
+			}
+			
+			if (width & 2) {
+				fb_writew(bgx,p); p += 2;
+			}
+			if (width & 1)
+				fb_writeb(bgx,(u8*)p);
+			break;
+
+		}		
+		dst += dstbytes;
+	}
+}
+
+void fbsplash_copy(u8 *dst, u8 *src, int height, int width, int linebytes, 
+		   int srclinebytes, int bpp)
+{
+	int i;
+
+	while (height-- > 0) {
+		u32 *p = (u32 *)dst;
+		u32 *q = (u32 *)src;
+
+		switch (bpp) {
+	
+		case 32:
+			for (i=0; i < width; i++)
+				fb_writel(*q++, p++);
+			break;	
+		case 24:	
+			for (i=0; i < (width*3/4); i++)
+				fb_writel(*q++, p++);
+			if ((width*3) % 4) {
+				if (width & 2) {
+					fb_writeb(*(u8*)q, (u8*)p);
+				} else if (width & 1) {
+					fb_writew(*(u16*)q, (u16*)p);
+					fb_writeb(*(u8*)((u16*)q+1),(u8*)((u16*)p+2));
+				}
+			}
+			break;
+		case 16:
+			for (i=0; i < width/4; i++) {
+				fb_writel(*q++, p++);
+				fb_writel(*q++, p++);
+			}
+			if (width & 2)
+				fb_writel(*q++, p++);
+			if (width & 1)
+				fb_writew(*(u16*)q, (u16*)p);
+			break;
+		case 8:
+			for (i=0; i < width/4; i++) 
+				fb_writel(*q++, p++);
+				
+			if (width & 2) {
+				fb_writew(*(u16*)q, (u16*)p); 
+				q = (u32*) ((u16*)q + 1);
+				p = (u32*) ((u16*)p + 1);
+			}
+			if (width & 1)
+				fb_writeb(*(u8*)q, (u8*)p);
+			break;
+		}
+
+		dst += linebytes;
+		src += srclinebytes;
+	}
+}
+
+static void splashfill(struct fb_info *info, int sy, int sx, int height, 
+		       int width) 
+{
+	int bytespp = ((info->var.bits_per_pixel + 7) >> 3);
+	int d  = sy * info->fix.line_length + sx * bytespp;
+	int ds = (sy * info->var.xres + sx) * bytespp;
+
+	fbsplash_copy((u8 *)(info->screen_base + d), (u8 *)(info->splash.data + ds),
+		    height, width, info->fix.line_length, info->var.xres * bytespp,
+		    info->var.bits_per_pixel);
+}
+
+void fbsplash_clear(struct vc_data *vc, struct fb_info *info, int sy, int sx, 
+		    int height, int width)
+{
+	int bgshift = (vc->vc_hi_font_mask) ? 13 : 12;
+	int bg_color = attr_bgcol_ec(bgshift, vc);
+	int transparent = vc->vc_splash.bg_color == bg_color;
+	u8 *dst;
+
+	sy = sy * vc->vc_font.height + vc->vc_splash.ty;
+	sx = sx * vc->vc_font.width + vc->vc_splash.tx;
+	height *= vc->vc_font.height;
+	width *= vc->vc_font.width;
+
+	if (transparent) {
+		splashfill(info, sy, sx, height, width);
+	} else {
+		dst = (u8 *)(info->screen_base + sy * info->fix.line_length + 
+			     sx * ((info->var.bits_per_pixel + 7) >> 3));
+		splashset(dst, height, width, info->fix.line_length, cc2cx(bg_color), 
+			  info->var.bits_per_pixel);
+	}
+}
+
+void fbsplash_clear_margins(struct vc_data *vc, struct fb_info *info, 
+			    int bottom_only)
+{
+	unsigned int tw = vc->vc_cols*vc->vc_font.width;
+	unsigned int th = vc->vc_rows*vc->vc_font.height;
+
+	if (!bottom_only) {
+		/* top margin */
+		splashfill(info, 0, 0, vc->vc_splash.ty, info->var.xres);
+		/* left margin */
+		splashfill(info, vc->vc_splash.ty, 0, th, vc->vc_splash.tx);
+		/* right margin */
+		splashfill(info, vc->vc_splash.ty, vc->vc_splash.tx + tw, th, 
+			   info->var.xres - vc->vc_splash.tx - tw);
+	}
+	splashfill(info, vc->vc_splash.ty + th, 0, 
+		   info->var.yres - vc->vc_splash.ty - th, info->var.xres);
+}
+
+void fbsplash_bmove_redraw(struct vc_data *vc, struct fb_info *info, int y, 
+			   int sx, int dx, int width)
+{
+	u16 *d = (u16 *) (vc->vc_origin + vc->vc_size_row * y + dx * 2);
+	u16 *s = d + (dx - sx);
+	u16 *start = d;
+	u16 *ls = d;
+	u16 *le = d + width;
+	u16 c;
+	int x = dx;
+	u16 attr = 1;
+
+	do {
+		c = scr_readw(d);
+		if (attr != (c & 0xff00)) {
+			attr = c & 0xff00;
+			if (d > start) {
+				fbsplash_putcs(vc, info, start, d - start, y, x);
+				x += d - start;
+				start = d;
+			}
+		}
+		if (s >= ls && s < le && c == scr_readw(s)) {
+			if (d > start) {
+				fbsplash_putcs(vc, info, start, d - start, y, x);
+				x += d - start + 1;
+				start = d + 1;
+			} else {
+				x++;
+				start++;
+			}
+		}
+		s++;
+		d++;
+	} while (d < le);
+	if (d > start)
+		fbsplash_putcs(vc, info, start, d - start, y, x);
+}
+
+void fbsplash_blank(struct vc_data *vc, struct fb_info *info, int blank)
+{
+	if (blank) {
+		splashset((u8 *)info->screen_base, info->var.yres, info->var.xres,
+			  info->fix.line_length, 0, info->var.bits_per_pixel);
+	} else {
+		update_screen(vc->vc_num);
+		fbsplash_clear_margins(vc, info, 0);
+	}
+}
+
diff -Nru a/drivers/video/console/bitblit.c b/drivers/video/console/bitblit.c
--- a/drivers/video/console/bitblit.c	2005-02-13 14:35:45 +01:00
+++ b/drivers/video/console/bitblit.c	2005-02-13 14:35:45 +01:00
@@ -18,6 +18,7 @@
 #include <linux/console.h>
 #include <asm/types.h>
 #include "fbcon.h"
+#include "../fbsplash.h"
 
 /*
  * Accelerated handlers.
@@ -84,6 +85,13 @@
 	area.height = height * vc->vc_font.height;
 	area.width = width * vc->vc_font.width;
 
+	if (fbsplash_active(info, vc)) {
+ 		area.sx += vc->vc_splash.tx;
+ 		area.sy += vc->vc_splash.ty;
+ 		area.dx += vc->vc_splash.tx;
+ 		area.dy += vc->vc_splash.ty;
+ 	}
+
 	info->fbops->fb_copyarea(info, &area);
 }
 
@@ -373,7 +381,11 @@
 	cursor.image.depth = 1;
 	cursor.rop = ROP_XOR;
 
-	info->fbops->fb_cursor(info, &cursor);
+	if (fbsplash_active(info, vc)) {
+		fbsplash_cursor(info, &cursor);
+	} else {
+		info->fbops->fb_cursor(info, &cursor);
+	}
 
 	ops->cursor_reset = 0;
 }
diff -Nru a/drivers/video/console/fbcon.c b/drivers/video/console/fbcon.c
--- a/drivers/video/console/fbcon.c	2005-02-13 14:35:45 +01:00
+++ b/drivers/video/console/fbcon.c	2005-02-13 14:35:45 +01:00
@@ -93,6 +93,7 @@
 #endif
 
 #include "fbcon.h"
+#include "../fbsplash.h"
 
 #ifdef FBCONDEBUG
 #  define DPRINTK(fmt, args...) printk(KERN_DEBUG "%s: " fmt, __FUNCTION__ , ## args)
@@ -206,7 +207,7 @@
 		vt_cons[vc->vc_num]->vc_mode != KD_TEXT);
 }
 
-static inline int get_color(struct vc_data *vc, struct fb_info *info,
+inline int get_color(struct vc_data *vc, struct fb_info *info,
 	      u16 c, int is_fg)
 {
 	int depth = fb_get_color_depth(info);
@@ -279,6 +280,7 @@
 		CM_ERASE : CM_DRAW;
 	ops->cursor(vc, info, p, mode, get_color(vc, info, c, 1),
 		    get_color(vc, info, c, 0));
+	
 	release_console_sem();
 }
 
@@ -404,6 +406,8 @@
 		info_idx = -1;
 	}
 
+	fbsplash_init();
+
 	return err;
 }
 
@@ -802,6 +806,12 @@
 
 	cols = info->var.xres / vc->vc_font.width;
 	rows = info->var.yres / vc->vc_font.height;
+
+	if (fbsplash_active(info, vc)) {
+		cols = vc->vc_splash.twidth / vc->vc_font.width;
+		rows = vc->vc_splash.theight / vc->vc_font.height;
+	}
+
 	vc_resize(vc->vc_num, cols, rows);
 
 	DPRINTK("mode:   %s\n", info->fix.id);
@@ -897,7 +907,7 @@
 	if (info_idx == -1 || info == NULL)
 	    return;
 	if (vc->vc_num != display_fg || logo_shown == FBCON_LOGO_DONTSHOW ||
-	    (info->fix.type == FB_TYPE_TEXT))
+	    (info->fix.type == FB_TYPE_TEXT) || fbsplash_active(info, vc))
 		logo = 0;
 
 	info->var.xoffset = info->var.yoffset = p->yscroll = 0;	/* reset wrap/pan */
@@ -1038,6 +1048,11 @@
 	if (!height || !width)
 		return;
 
+ 	if (fbsplash_active(info, vc)) {
+ 		fbsplash_clear(vc, info, sy, sx, height, width);
+ 		return;
+ 	}
+ 	
 	/* Split blits that cross physical y_wrap boundary */
 
 	y_break = p->vrows - p->yscroll;
@@ -1057,10 +1072,15 @@
 	struct display *p = &fb_display[vc->vc_num];
 	struct fbcon_ops *ops = info->fbcon_par;
 
-	if (!fbcon_is_inactive(vc, info))
-		ops->putcs(vc, info, s, count, real_y(p, ypos), xpos,
-			   get_color(vc, info, scr_readw(s), 1),
-			   get_color(vc, info, scr_readw(s), 0));
+	if (!fbcon_is_inactive(vc, info)) {
+		
+		if (fbsplash_active(info, vc))
+			fbsplash_putcs(vc, info, s, count, ypos, xpos);
+		else
+			ops->putcs(vc, info, s, count, real_y(p, ypos), xpos,
+				   get_color(vc, info, scr_readw(s), 1),
+				   get_color(vc, info, scr_readw(s), 0));
+	}
 }
 
 static void fbcon_putc(struct vc_data *vc, int c, int ypos, int xpos)
@@ -1076,8 +1096,15 @@
 	struct fb_info *info = registered_fb[con2fb_map[vc->vc_num]];
 	struct fbcon_ops *ops = info->fbcon_par;
 
-	if (!fbcon_is_inactive(vc, info))
-		ops->clear_margins(vc, info, bottom_only);
+	if (!fbcon_is_inactive(vc, info)) {
+	 	if (fbsplash_active(info, vc)) {
+ 			if (!fbsplash_isverbose)
+ 				return;
+	 		fbsplash_clear_margins(vc, info, bottom_only);
+ 		} else {
+			ops->clear_margins(vc, info, bottom_only);
+		}
+	}
 }
 
 static void fbcon_cursor(struct vc_data *vc, int mode)
@@ -1544,7 +1571,7 @@
 	int scroll_partial = info->flags & FBINFO_PARTIAL_PAN_OK;
 
 	if (fbcon_is_inactive(vc, info))
-		return -EINVAL;
+		return (fbsplash_issilent) ? 0 : -EINVAL;
 
 	fbcon_cursor(vc, CM_ERASE);
 
@@ -1560,7 +1587,7 @@
 			count = vc->vc_rows;
 		if (softback_top)
 			fbcon_softback_note(vc, t, count);
-		if (logo_shown >= 0)
+		if (logo_shown >= 0 || fbsplash_active(info, vc))
 			goto redraw_up;
 		switch (p->scrollmode) {
 		case SCROLL_MOVE:
@@ -1646,6 +1673,8 @@
 	case SM_DOWN:
 		if (count > vc->vc_rows)	/* Maximum realistic size */
 			count = vc->vc_rows;
+		if (fbsplash_active(info, vc))
+			goto redraw_down;
 		switch (p->scrollmode) {
 		case SCROLL_MOVE:
 			ops->bmove(vc, info, t, 0, t + count, 0,
@@ -1788,6 +1817,13 @@
 		}
 		return;
 	}
+
+	if (fbsplash_active(info, vc) && sy == dy && height == 1) {
+ 		/* must use slower redraw bmove to keep background pic intact */
+ 		fbsplash_bmove_redraw(vc, info, sy, sx, dx, width);
+ 		return;
+ 	}
+	
 	ops->bmove(vc, info, real_y(p, sy), sx, real_y(p, dy), dx,
 		   height, width);
 }
@@ -1842,7 +1878,8 @@
 	var.yres = height * fh;
 	x_diff = info->var.xres - var.xres;
 	y_diff = info->var.yres - var.yres;
-	if (x_diff < 0 || x_diff > fw || (y_diff < 0 || y_diff > fh)) {
+
+	if ((x_diff < 0 || x_diff > fw || (y_diff < 0 || y_diff > fh)) && !vc->vc_splash.state) {
 		struct fb_videomode *mode;
 
 		DPRINTK("attempting resize %ix%i\n", var.xres, var.yres);
@@ -1889,7 +1926,17 @@
 	int i, prev_console, do_set_par = 0;
 
 	info = registered_fb[con2fb_map[vc->vc_num]];
+	prev_console = ((struct fbcon_ops *)info->fbcon_par)->currcon;
+	
+	if (fbsplash_active_vc(vc)) {
+		struct vc_data *vc_curr = vc_cons[prev_console].d;
 
+		if (!vc_curr->vc_splash.theme || strcmp(vc->vc_splash.theme, vc_curr->vc_splash.theme)) {
+			if (fbsplash_call_helper("getpic", vc->vc_num))
+				fbsplash_disable(vc, 0);
+		}
+	}
+	
 	if (softback_top) {
 		int l = fbcon_softback_size / vc->vc_size_row;
 		if (softback_lines)
@@ -1915,8 +1962,6 @@
 		logo_shown = FBCON_LOGO_CANSHOW;
 	}
 
-	prev_console = ((struct fbcon_ops *)info->fbcon_par)->currcon;
-
 	/*
 	 * FIXME: If we have multiple fbdev's loaded, we need to
 	 * update all info->currcon.  Perhaps, we can place this
@@ -2043,7 +2088,9 @@
 			fbcon_cursor(vc, blank ? CM_ERASE : CM_DRAW);
 			ops->cursor_flash = (!blank);
 
-			if (fb_blank(info, blank))
+ 			if (fbsplash_active(info, vc))
+				fbsplash_blank(vc, info, blank);
+			else
 				fbcon_generic_blank(vc, info, blank);
 		}
 
@@ -2196,9 +2243,15 @@
 	}
 
 	if (resize) {
-		/* reset wrap/pan */
-		info->var.xoffset = info->var.yoffset = p->yscroll = 0;
-		vc_resize(vc->vc_num, info->var.xres / w, info->var.yres / h);
+		u32 xres = info->var.xres, yres = info->var.yres;
+  		/* reset wrap/pan */
+  		info->var.xoffset = info->var.yoffset = p->yscroll = 0;
+		if (fbsplash_active(info, vc)) {
+			xres = vc->vc_splash.twidth;
+			yres = vc->vc_splash.theight;
+		}
+		vc_resize(vc->vc_num, xres / w, yres / h);
+		
 		if (CON_IS_VISIBLE(vc) && softback_buf) {
 			int l = fbcon_softback_size / vc->vc_size_row;
 			if (l > 5)
@@ -2348,7 +2401,49 @@
 	} else
 		fb_copy_cmap(fb_default_cmap(1 << depth), &palette_cmap);
 
-	return fb_set_cmap(&palette_cmap, info);
+	if (fbsplash_active(info, vc_cons[fg_console].d) &&
+	    info->fix.visual == FB_VISUAL_DIRECTCOLOR) {
+
+		u16 *red, *green, *blue;
+		int minlen = min(min(info->var.red.length, info->var.green.length), 
+				     info->var.blue.length);
+		int h;
+
+		struct fb_cmap cmap = {
+			.start = 0,
+			.len = (1 << minlen),
+			.red = NULL,
+			.green = NULL,
+			.blue = NULL,
+			.transp = NULL
+		};
+
+		red = kmalloc(256 * sizeof(u16) * 3, GFP_KERNEL);
+	
+		if (!red)
+			goto out;		
+	
+		green = red + 256;
+		blue = green + 256;
+		cmap.red = red;
+		cmap.green = green;
+		cmap.blue = blue;
+		
+		for (i = 0; i < cmap.len; i++) {
+			red[i] = green[i] = blue[i] = (0xffff * i)/(cmap.len-1);
+		}
+
+		h = fb_set_cmap(&cmap, info);
+		fbsplash_fix_pseudo_pal(info, vc_cons[fg_console].d);
+		kfree(red);
+		
+		return h;
+		
+	} else if (fbsplash_active(info, vc_cons[fg_console].d) && 
+		   info->var.bits_per_pixel == 8 && info->splash.cmap.red != NULL) 
+		fb_set_cmap(&info->splash.cmap, info);
+		
+out:	return fb_set_cmap(&palette_cmap, info);
 }
 
 static u16 *fbcon_screen_pos(struct vc_data *vc, int offset)
@@ -2569,7 +2664,14 @@
 		var_to_display(p, &info->var, info);
 		cols = info->var.xres / vc->vc_font.width;
 		rows = info->var.yres / vc->vc_font.height;
-		vc_resize(vc->vc_num, cols, rows);
+		
+		if (!fbsplash_active(info, vc)) {
+			vc_resize(vc->vc_num, cols, rows);
+		} else {
+			if (fbsplash_call_helper("modechange", vc->vc_num))
+				fbsplash_disable(vc, 0);
+		}
+	
 		updatescrollmode(p, info, vc);
 		scrollback_max = 0;
 		scrollback_current = 0;
diff -Nru a/drivers/video/fbcmap.c b/drivers/video/fbcmap.c
--- a/drivers/video/fbcmap.c	2005-02-13 14:35:45 +01:00
+++ b/drivers/video/fbcmap.c	2005-02-13 14:35:45 +01:00
@@ -16,6 +16,7 @@
 #include <linux/tty.h>
 #include <linux/fb.h>
 #include <linux/slab.h>
+#include "fbsplash.h"
 
 #include <asm/uaccess.h>
 
@@ -235,6 +236,10 @@
 					      info))
 			break;
 	}
+	fb_copy_cmap(cmap, &info->cmap);
+	if (fbsplash_active(info, vc_cons[fg_console].d) &&
+	    info->fix.visual == FB_VISUAL_DIRECTCOLOR) 
+		fbsplash_fix_pseudo_pal(info, vc_cons[fg_console].d);
 	return 0;
 }
 
@@ -265,6 +270,9 @@
 		if (transp)
 			transp++;
 	}
+	if (fbsplash_active(info, vc_cons[fg_console].d) &&
+	    info->fix.visual == FB_VISUAL_DIRECTCOLOR) 
+		fbsplash_fix_pseudo_pal(info, vc_cons[fg_console].d);
 	return 0;
 }
 
diff -Nru a/drivers/video/fbsplash.c b/drivers/video/fbsplash.c
--- /dev/null	Wed Dec 31 16:00:00 196900
+++ b/drivers/video/fbsplash.c	2005-02-13 14:35:45 +01:00
@@ -0,0 +1,534 @@
+/* 
+ *  linux/drivers/video/fbsplash.c -- Framebuffer splash routines
+ *
+ *  Copyright (C) 2004 Michal Januszewski <spock@gentoo.org>
+ *
+ *  Code based upon "Bootsplash" (C) 2001-2003 
+ *       Volker Poplawski <volker@poplawski.de>,
+ *       Stefan Reinauer <stepan@suse.de>,
+ *       Steffen Winterfeldt <snwint@suse.de>,
+ *       Michael Schroeder <mls@suse.de>,
+ *       Ken Wimer <wimer@suse.de>.
+ *
+ *  Splash render routines are located in /linux/drivers/video/cfbsplash.c
+ * 
+ *  This file is subject to the terms and conditions of the GNU General Public
+ *  License.  See the file COPYING in the main directory of this archive for
+ *  more details.
+ * 
+ */
+#include <linux/config.h>
+#include <linux/module.h>
+#include <linux/kernel.h>
+#include <linux/string.h>
+#include <linux/types.h>
+#include <linux/fb.h>
+#include <linux/vt_kern.h>
+#include <linux/vmalloc.h>
+#include <linux/unistd.h>
+#include <linux/syscalls.h>
+#include <linux/init.h>
+#include <linux/proc_fs.h>
+#include <linux/workqueue.h>
+#include <linux/kmod.h>
+#include <linux/miscdevice.h>
+#include <linux/device.h>
+#include <linux/fs.h>
+
+#include <asm/uaccess.h>
+#include <asm/irq.h>
+#include <asm/system.h>
+
+#include "console/fbcon.h"
+#include "fbsplash.h"
+
+#define SPLASH_VERSION "0.9"
+
+#ifdef DEBUG
+#define DPRINTK(fmt, args...)	printk(KERN_DEBUG "%s: " fmt, __FUNCTION__ , ## args)
+#else
+#define DPRINTK(fmt, args...)
+#endif
+
+extern signed char con2fb_map[];
+
+static void splash_work_verbose(void *data);
+static int fbsplash_enable(struct vc_data *vc);
+
+DECLARE_WORK(splash_wq_verbose, splash_work_verbose, NULL);
+
+int fbsplash_mode = 0;
+char fbsplash_path[KMOD_PATH_LEN] = "/sbin/splash_helper";
+static char fbsplash_theme[FB_SPLASH_THEME_LEN] = "default";
+
+int fbsplash_call_helper(char* cmd, unsigned short vc)
+{
+	char *envp[] = {
+		"HOME=/",
+		"PATH=/sbin:/bin",
+		NULL
+	};
+
+	char tfb[5];
+	char tcons[5];
+	unsigned char fb = (int) con2fb_map[vc];
+
+	char *argv[] = {
+		fbsplash_path,
+		"1",
+		cmd,
+		tcons,
+		tfb,
+		(fbsplash_mode == FB_SPLASH_MODE_SILENT) ? "s" : "v",
+		vc_cons[vc].d->vc_splash.theme,
+		NULL
+	};
+
+	snprintf(tfb,5,"%d",fb);
+	snprintf(tcons,5,"%d",vc);
+	
+	return call_usermodehelper(fbsplash_path, argv, envp, 1);
+}
+
+int fbsplash_verbose(void)
+{
+	if (fbsplash_mode == FB_SPLASH_MODE_VERBOSE)
+		return 0;
+
+	fbsplash_mode = FB_SPLASH_MODE_VERBOSE;
+	
+	/* We have to do the switch from a workqueue helper, because chances are this
+	 * function gets called from interrupt context, and we can't get hold of the 
+	 * console sem this way. */
+	schedule_work(&splash_wq_verbose);
+
+	return 1;
+}
+
+/* Switches the first console to verbose mode. */
+static void splash_work_verbose(void *data)
+{
+	struct fb_info *info;
+	struct vc_data *vc;
+	
+	vc = vc_cons[0].d;
+	info = registered_fb[(int) con2fb_map[vc->vc_num]];
+
+	if (info == NULL)
+		return;
+
+	printk(KERN_INFO "fbsplash: switching to verbose mode\n");
+
+	vt_cons[0]->vc_mode = KD_TEXT;
+
+	acquire_console_sem();
+	do_unblank_screen(0);
+
+	if (info->splash.data) {
+		fbsplash_enable(vc);
+	} else {
+		fbsplash_disable(vc, 1);
+	}
+	release_console_sem();
+}
+
+/* Disables fbsplash on a virtual console; called with console sem held. */
+int fbsplash_disable(struct vc_data *vc, unsigned char redraw)
+{
+	struct fb_info* info;
+
+	if (!vc->vc_splash.state)
+		return -EINVAL;
+
+	info = registered_fb[(int) con2fb_map[vc->vc_num]];
+
+	if (info == NULL)
+		return -EINVAL;
+
+	vc->vc_splash.state = 0; 
+	vc_resize(vc->vc_num, info->var.xres / vc->vc_font.width, 
+		  info->var.yres / vc->vc_font.height);
+
+	if (fg_console == vc->vc_num && redraw) {
+		redraw_screen(fg_console, 0);
+		update_region(fg_console, vc->vc_origin + 
+			      vc->vc_size_row * vc->vc_top, 
+			      vc->vc_size_row * (vc->vc_bottom - vc->vc_top) / 2);
+	}
+
+	printk(KERN_INFO "fbsplash: switched splash state to 'off' on console %d\n", 
+			 vc->vc_num);
+
+	return 0;
+}
+
+/* Enables fbsplash on a virtual console; called with console sem held. */
+static int fbsplash_enable(struct vc_data *vc)
+{
+	struct fb_info* info;
+
+	info = registered_fb[(int) con2fb_map[vc->vc_num]];
+		
+	if (vc->vc_splash.twidth == 0 || vc->vc_splash.theight == 0 || 
+	    info == NULL || !info->splash.data || vc->vc_splash.state)
+		return -EINVAL;
+	
+	vc->vc_splash.state = 1;
+	vc_resize(vc->vc_num, vc->vc_splash.twidth / vc->vc_font.width, 
+		  vc->vc_splash.theight / vc->vc_font.height);
+
+	if (fg_console == vc->vc_num) {
+		redraw_screen(fg_console, 0);
+		update_region(fg_console, vc->vc_origin + 
+			      vc->vc_size_row * vc->vc_top, 
+			      vc->vc_size_row * (vc->vc_bottom - vc->vc_top) / 2);
+		fbsplash_clear_margins(vc, info, 0);
+	}
+
+	printk(KERN_INFO "fbsplash: switched splash state to 'on' on console %d\n", 
+			 vc->vc_num);
+
+	return 0;
+}
+
+static int __init splash_setup(char *options)
+{
+	char *this_opt;
+	
+	while ((this_opt = strsep(&options, ",")) != NULL) {
+	
+		if (!strcmp(this_opt, "silent")) {
+			fbsplash_mode = FB_SPLASH_MODE_SILENT;
+			printk(KERN_INFO "fbsplash: silent\n");
+		} else if (!strcmp(this_opt, "verbose")) {
+			fbsplash_mode = FB_SPLASH_MODE_VERBOSE;
+			printk(KERN_INFO "fbsplash: verbose\n");
+		} else if (!strcmp(this_opt, "off")) {
+			fbsplash_mode = 0;
+		} else if (!strncmp(this_opt, "theme:", 6)) {
+			strncpy(fbsplash_theme, this_opt+6, 64);
+			printk(KERN_INFO "fbsplash: theme %s\n", fbsplash_theme);
+		} else {
+			printk(KERN_WARNING "fbsplash: unrecognized option %s\n", this_opt);
+		}
+	}
+	return 0;
+}
+
+__setup("splash=", splash_setup);
+
+static int splash_get_info(char *buf, char **start, off_t fpos, int length)
+{
+	char *p = buf;
+	
+	p += sprintf(p, "Framebuffer splash v%s, mode: %s\n", SPLASH_VERSION,
+		     ((fbsplash_mode == 2) ? "silent" : (fbsplash_mode == 1) ? "verbose" : "off"));
+
+	return p - buf;
+}
+
+static inline int fbsplash_ioctl_dosetstate(struct vc_data *vc, unsigned int __user* state, unsigned char origin)
+{
+	int tmp, ret;
+
+	if (get_user(tmp, state))
+		return -EFAULT;
+
+	if (origin == FB_SPLASH_IO_ORIG_USER)
+		acquire_console_sem();
+	if (!tmp)
+		ret = fbsplash_disable(vc, 1);
+	else
+		ret = fbsplash_enable(vc);
+	if (origin == FB_SPLASH_IO_ORIG_USER)
+		release_console_sem();
+
+	return ret;
+}
+
+static inline int fbsplash_ioctl_dogetstate(struct vc_data *vc, unsigned int __user *state)
+{
+	return put_user(vc->vc_splash.state, (unsigned int __user*) state);
+}
+
+static int fbsplash_ioctl_dosetcfg(struct vc_data *vc, struct vc_splash __user *arg, unsigned char origin)
+{
+	struct vc_splash cfg;
+	struct fb_info *info;
+	int len;
+	char *tmp;
+	
+	info = registered_fb[(int) con2fb_map[vc->vc_num]];
+
+	if (copy_from_user(&cfg, arg, sizeof(struct vc_splash)))
+		return -EFAULT;
+	if (info == NULL || !cfg.twidth || !cfg.theight || 
+	    cfg.tx + cfg.twidth  > info->var.xres ||
+	    cfg.ty + cfg.theight > info->var.yres)
+		return -EINVAL;
+
+	len = strlen_user(cfg.theme);
+	if (!len || len > FB_SPLASH_THEME_LEN)
+		return -EINVAL;
+	tmp = kmalloc(len, GFP_KERNEL);
+	if (!tmp)
+		return -ENOMEM;
+	if (copy_from_user(tmp, (void __user *)cfg.theme, len))
+		return -EFAULT;
+	cfg.theme = tmp;
+	cfg.state = 0;
+
+	/* If this ioctl is a response to a request from kernel, the console sem
+	 * is already held; we also don't need to disable splash because either the
+	 * new config and background picture will be successfully loaded, and the 
+	 * splash will stay on, or in case of a failure it'll be turned off in fbcon. */
+	if (origin == FB_SPLASH_IO_ORIG_USER) {
+		acquire_console_sem();
+		if (vc->vc_splash.state)
+			fbsplash_disable(vc, 1);
+	}
+
+	if (vc->vc_splash.theme)
+		kfree(vc->vc_splash.theme);
+
+	vc->vc_splash = cfg;
+
+	if (origin == FB_SPLASH_IO_ORIG_USER)
+		release_console_sem();
+
+	printk(KERN_INFO "fbsplash: console %d using theme '%s'\n", 
+			 vc->vc_num, vc->vc_splash.theme);
+	return 0;	
+}
+
+static int fbsplash_ioctl_dogetcfg(struct vc_data *vc, struct vc_splash __user *arg)
+{
+	struct vc_splash splash;
+	char __user *tmp;
+
+	if (get_user(tmp, &arg->theme))
+		return -EFAULT;
+	
+	splash = vc->vc_splash;
+	splash.theme = tmp;
+
+	if (vc->vc_splash.theme) {
+		if (copy_to_user(tmp, vc->vc_splash.theme, strlen(vc->vc_splash.theme) + 1))
+			return -EFAULT;
+	} else
+		if (put_user(0, tmp))
+			return -EFAULT;
+
+	if (copy_to_user(arg, &splash, sizeof(struct vc_splash)))
+		return -EFAULT;
+
+	return 0;
+}
+
+static int fbsplash_ioctl_dosetpic(struct vc_data *vc, struct fb_image __user *arg, unsigned char origin)
+{
+	struct fb_image img;
+	struct fb_info *info;
+	int len;
+	u8 *tmp;
+	
+	if (vc->vc_num != fg_console) 
+		return -EINVAL;
+
+	info = registered_fb[(int) con2fb_map[vc->vc_num]];
+	
+	if (info == NULL)
+		return -EINVAL;
+	
+	if (copy_from_user(&img, arg, sizeof(struct fb_image)))
+		return -EFAULT;
+	
+	if (img.width != info->var.xres || img.height != info->var.yres) {
+		printk(KERN_ERR "fbsplash: picture dimensions mismatch\n");
+		return -EINVAL;
+	}
+
+	if (img.depth != info->var.bits_per_pixel) {
+		printk(KERN_ERR "fbsplash: picture depth mismatch\n");
+		return -EINVAL;
+	}
+		
+	if (img.depth == 8) {
+		if (!img.cmap.len || !img.cmap.red || !img.cmap.green || 
+		    !img.cmap.blue)
+			return -EINVAL;
+		
+		tmp = vmalloc(img.cmap.len * 3 * 2);
+		if (!tmp)
+			return -ENOMEM;
+
+		if (copy_from_user(tmp, (void __user*)img.cmap.red, img.cmap.len * 2) ||
+		    copy_from_user(tmp + (img.cmap.len << 1),
+			    	   (void __user*)img.cmap.green, (img.cmap.len << 1)) ||
+		    copy_from_user(tmp + (img.cmap.len << 2),
+			    	   (void __user*)img.cmap.blue, (img.cmap.len << 1))) {
+			vfree(tmp);
+			return -EFAULT;
+		}
+			
+		img.cmap.transp = NULL;
+		img.cmap.red = (u16*)tmp;
+		img.cmap.green = img.cmap.red + img.cmap.len;
+		img.cmap.blue = img.cmap.green + img.cmap.len;
+	} else {
+		img.cmap.red = NULL;
+	}
+		
+	len = ((img.depth + 7) >> 3) * img.width * img.height;
+	tmp = vmalloc(len);
+
+	if (!tmp)
+		goto out;
+
+	if (copy_from_user(tmp, (void __user*)img.data, len))
+		goto out;
+		
+	img.data = tmp;
+
+	/* If this ioctl is a response to a request from kernel, the console sem
+	 * is already held. */
+	if (origin == FB_SPLASH_IO_ORIG_USER)
+		acquire_console_sem();
+	
+	if (info->splash.data)
+		vfree((u8*)info->splash.data);
+	if (info->splash.cmap.red)
+		vfree(info->splash.cmap.red);
+	
+	info->splash = img;
+
+	if (origin == FB_SPLASH_IO_ORIG_USER)
+		release_console_sem();
+
+	return 0;
+
+out:	if (img.cmap.red)
+		vfree(img.cmap.red);
+	if (tmp)
+		vfree(tmp);
+	return -ENOMEM;
+}
+
+static int splash_ioctl(struct inode * inode, struct file *filp, u_int cmd, 
+			u_long arg)
+{
+	struct fb_splash_iowrapper __user *wrapper = (void __user*) arg;
+	struct vc_data *vc = NULL;
+	unsigned short vc_num = 0;
+	unsigned char origin = 0;
+	void __user *data = NULL;
+	
+	if (cmd != FBIOSPLASH_SETMODE && cmd != FBIOSPLASH_GETMODE)
+	{
+		if (verify_area(VERIFY_READ, wrapper, 
+				sizeof(struct fb_splash_iowrapper)))
+			return -EFAULT;
+	
+		__get_user(vc_num, &wrapper->vc);
+		__get_user(origin, &wrapper->origin);
+		__get_user(data, &wrapper->data);
+		
+		if (!vc_cons_allocated(vc_num))
+			return -EINVAL;
+
+		vc = vc_cons[vc_num].d;
+	}
+	
+	switch (cmd) {
+	
+	case FBIOSPLASH_SETMODE:
+		if (arg != FB_SPLASH_MODE_SILENT && arg != FB_SPLASH_MODE_VERBOSE)
+			return -EINVAL;
+	
+		if (fbsplash_mode == arg)
+			return 0;
+	
+		fbsplash_mode = arg;
+		if (arg == FB_SPLASH_MODE_VERBOSE)
+			splash_work_verbose(NULL);
+		return 0;
+	
+	case FBIOSPLASH_GETMODE:
+		return put_user(fbsplash_mode, (unsigned int __user *)arg);
+	case FBIOSPLASH_SETPIC:
+		return fbsplash_ioctl_dosetpic(vc, (struct fb_image __user*)data, origin);
+	case FBIOSPLASH_SETCFG:
+		return fbsplash_ioctl_dosetcfg(vc, (struct vc_splash*)data, origin);
+	case FBIOSPLASH_GETCFG:
+		return fbsplash_ioctl_dogetcfg(vc, (struct vc_splash*)data);
+	case FBIOSPLASH_SETSTATE:
+		return fbsplash_ioctl_dosetstate(vc, (unsigned int *)data, origin);
+	case FBIOSPLASH_GETSTATE:
+		return fbsplash_ioctl_dogetstate(vc, (unsigned int *)data);
+	default:
+		return -ENOIOCTLCMD;
+	}	
+}
+
+static struct file_operations splash_ops = {
+	.owner = THIS_MODULE,
+	.ioctl = splash_ioctl
+};
+
+static struct miscdevice splash_dev = {
+	.minor = MISC_DYNAMIC_MINOR,
+	.name = "fbsplash",
+	.fops = &splash_ops
+};
+
+int fbsplash_init(void)
+{
+	struct proc_dir_entry *splash_proc;
+	struct fb_info *info;
+	struct vc_data *vc;
+	int i;
+	
+	vc = vc_cons[0].d;
+	info = registered_fb[0];
+
+	for (i = 0; i < num_registered_fb; i++) {
+		registered_fb[i]->splash.data = NULL;
+		registered_fb[i]->splash.cmap.red = NULL;
+	}
+
+	for (i = 0; i < MAX_NR_CONSOLES && vc_cons[i].d; i++) {
+		vc_cons[i].d->vc_splash.state = vc_cons[i].d->vc_splash.twidth = 
+						vc_cons[i].d->vc_splash.theight = 0;
+		vc_cons[i].d->vc_splash.theme = NULL;
+	}
+
+	i = misc_register(&splash_dev);
+	if (i) {
+		printk(KERN_ERR "fbsplash: failed to register device\n");
+		return i;
+	}
+
+#ifdef CONFIG_PROC_FS
+	splash_proc = create_proc_info_entry("fbsplash", 0, NULL, splash_get_info);
+	if (splash_proc)
+		splash_proc->owner = THIS_MODULE;
+#endif
+	
+	if (fbsplash_mode && info) {
+		vc->vc_splash.theme = kmalloc((strlen(fbsplash_theme)+1) * sizeof(char), GFP_KERNEL);
+		if (!vc->vc_splash.theme) {
+			misc_deregister(&splash_dev);
+			return -ENOMEM;
+		}
+		
+		strcpy(vc->vc_splash.theme, fbsplash_theme);
+		if (fbsplash_call_helper("init", 0))
+			fbsplash_mode = FB_SPLASH_MODE_VERBOSE;
+	} else {
+		fbsplash_mode = FB_SPLASH_MODE_VERBOSE;
+	}
+		
+	return 0;
+}
+
+EXPORT_SYMBOL(fbsplash_path);
diff -Nru a/drivers/video/fbsplash.h b/drivers/video/fbsplash.h
--- /dev/null	Wed Dec 31 16:00:00 196900
+++ b/drivers/video/fbsplash.h	2005-02-13 14:35:45 +01:00
@@ -0,0 +1,78 @@
+/* 
+ *  linux/drivers/video/fbsplash.h -- Framebuffer splash headers
+ *
+ *  Copyright (C) 2004 Michal Januszewski <spock@gentoo.org>
+ *
+ */
+
+#ifndef __FB_SPLASH_H
+#define __FB_SPLASH_H
+
+#ifndef _LINUX_FB_H
+#include <linux/fb.h>
+#endif
+
+/* This is needed for vc_cons in fbcmap.c */
+#include <linux/vt_kern.h>
+
+struct fb_cursor;
+struct fb_info;
+struct vc_data;
+
+#ifdef CONFIG_FB_SPLASH
+/* fbsplash.c */
+int fbsplash_init(void);
+int fbsplash_verbose(void);
+int fbsplash_call_helper(char* cmd, unsigned short cons);
+int fbsplash_disable(struct vc_data *vc, unsigned char redraw);
+
+/* cfbsplash.c */
+void fbsplash_putcs(struct vc_data *vc, struct fb_info *info, const unsigned short *s, int count, int yy, int xx);
+void fbsplash_cursor(struct fb_info *info, struct fb_cursor *cursor);
+void fbsplash_clear(struct vc_data *vc, struct fb_info *info, int sy, int sx, int height, int width);
+void fbsplash_clear_margins(struct vc_data *vc, struct fb_info *info, int bottom_only);
+void fbsplash_blank(struct vc_data *vc, struct fb_info *info, int blank);
+void fbsplash_bmove_redraw(struct vc_data *vc, struct fb_info *info, int y, int sx, int dx, int width);
+void fbsplash_copy(u8 *dst, u8 *src, int height, int width, int linebytes, int srclinesbytes, int bpp);
+void fbsplash_fix_pseudo_pal(struct fb_info *info, struct vc_data *vc);
+
+/* vt.c */
+void acquire_console_sem(void);
+void release_console_sem(void);
+void do_unblank_screen(int entering_gfx);
+
+extern int fbsplash_mode;
+#define fbsplash_isverbose (fbsplash_mode == FB_SPLASH_MODE_VERBOSE)
+#define fbsplash_issilent  (fbsplash_mode == FB_SPLASH_MODE_SILENT)
+
+/* struct vc_data *y */
+#define fbsplash_active_vc(y) (y->vc_splash.state && y->vc_splash.theme) 
+
+/* struct fb_info *x, struct vc_data *y */
+#define fbsplash_active(x,y) (x->splash.data && fbsplash_active_vc(y) &&		\
+			      x->splash.width == x->var.xres && 			\
+			      x->splash.height == x->var.yres && 			\
+			      x->splash.depth == x->var.bits_per_pixel)
+
+#else /* CONFIG_FB_SPLASH */
+
+static inline void fbsplash_putcs(struct vc_data *vc, struct fb_info *info, const unsigned short *s, int count, int yy, int xx) {}
+static inline void fbsplash_putc(struct vc_data *vc, struct fb_info *info, int c, int ypos, int xpos) {}
+static inline void fbsplash_cursor(struct fb_info *info, struct fb_cursor *cursor) {}
+static inline void fbsplash_clear(struct vc_data *vc, struct fb_info *info, int sy, int sx, int height, int width) {}
+static inline void fbsplash_clear_margins(struct vc_data *vc, struct fb_info *info, int bottom_only) {}
+static inline void fbsplash_blank(struct vc_data *vc, struct fb_info *info, int blank) {}
+static inline void fbsplash_bmove_redraw(struct vc_data *vc, struct fb_info *info, int y, int sx, int dx, int width) {}
+static inline int fbsplash_call_helper(char* cmd, unsigned short cons) { return 0; }
+static inline int fbsplash_init(void) { return 0; }
+static inline int fbsplash_verbose(void) { return 0; }
+static inline int fbsplash_disable(struct vc_data *vc, unsigned char redraw) { return 0; }
+
+#define fbsplash_issilent (0)
+#define fbsplash_isverbose (1)
+#define fbsplash_active_vc(y) (0)
+#define fbsplash_active(x,y) (0)
+
+#endif /* CONFIG_FB_SPLASH */
+
+#endif /* __FB_SPLASH_H */
diff -Nru a/include/linux/console_splash.h b/include/linux/console_splash.h
--- /dev/null	Wed Dec 31 16:00:00 196900
+++ b/include/linux/console_splash.h	2005-02-13 14:35:45 +01:00
@@ -0,0 +1,13 @@
+#ifndef _LINUX_CONSOLE_SPLASH_H_
+#define _LINUX_CONSOLE_SPLASH_H_ 1
+
+/* A structure used by the framebuffer splash code (drivers/video/fbsplash.c) */
+struct vc_splash {
+	__u8 bg_color;				/* The color that is to be treated as transparent */
+	__u8 state;				/* Current splash state: 0 = off, 1 = on */
+	__u16 tx, ty;				/* Top left corner coordinates of the text field */
+	__u16 twidth, theight;			/* Width and height of the text field */
+	char* theme;
+};
+
+#endif
diff -Nru a/include/linux/console_struct.h b/include/linux/console_struct.h
--- a/include/linux/console_struct.h	2005-02-13 14:35:45 +01:00
+++ b/include/linux/console_struct.h	2005-02-13 14:35:45 +01:00
@@ -12,6 +12,7 @@
 struct vt_struct;
 
 #define NPAR 16
+#include <linux/console_splash.h>
 
 struct vc_data {
 	unsigned short	vc_num;			/* Console number */
@@ -90,6 +91,8 @@
 	unsigned long	vc_uni_pagedir;
 	unsigned long	*vc_uni_pagedir_loc;  /* [!] Location of uni_pagedir variable for this console */
 	struct vt_struct *vc_vt;
+
+	struct vc_splash vc_splash;
 	/* additional information is in vt_kern.h */
 };
 
diff -Nru a/include/linux/fb.h b/include/linux/fb.h
--- a/include/linux/fb.h	2005-02-13 14:35:45 +01:00
+++ b/include/linux/fb.h	2005-02-13 14:35:45 +01:00
@@ -8,6 +8,13 @@
 #define FB_MAJOR		29
 #define FB_MAX			32	/* sufficient for now */
 
+struct fb_splash_iowrapper
+{
+	unsigned short vc;		/* Virtual console */
+	unsigned char origin;		/* Point of origin of the request */
+	void *data;
+};
+
 /* ioctls
    0x46 is 'F'								*/
 #define FBIOGET_VSCREENINFO	0x4600
@@ -35,7 +42,19 @@
 #define FBIOGET_HWCINFO         0x4616
 #define FBIOPUT_MODEINFO        0x4617
 #define FBIOGET_DISPINFO        0x4618
-
+#define FBIOSPLASH_SETCFG	_IOWR('F', 0x19, struct fb_splash_iowrapper)
+#define FBIOSPLASH_GETCFG	_IOR('F', 0x1A, struct fb_splash_iowrapper)
+#define FBIOSPLASH_SETSTATE	_IOWR('F', 0x1B, struct fb_splash_iowrapper)
+#define FBIOSPLASH_GETSTATE	_IOR('F', 0x1C, struct fb_splash_iowrapper)
+#define FBIOSPLASH_SETMODE	_IOWR('F', 0x1D, unsigned int)
+#define FBIOSPLASH_GETMODE	_IOR('F', 0x1E, unsigned int)
+#define FBIOSPLASH_SETPIC 	_IOWR('F', 0x1F, struct fb_splash_iowrapper)
+
+#define FB_SPLASH_MODE_VERBOSE	 	1	/* Verbose mode */
+#define FB_SPLASH_MODE_SILENT 		2	/* Silent mode */
+#define FB_SPLASH_THEME_LEN		128	/* Maximum lenght of a theme name */
+#define FB_SPLASH_IO_ORIG_KERNEL	0	/* Kernel ioctl origin */
+#define FB_SPLASH_IO_ORIG_USER		1 	/* User ioctl origin */
 
 #define FB_TYPE_PACKED_PIXELS		0	/* Packed Pixels	*/
 #define FB_TYPE_PLANES			1	/* Non interleaved planes */
@@ -724,6 +743,9 @@
 #define FBINFO_STATE_SUSPENDED	1
 	u32 state;			/* Hardware state i.e suspend */
 	void *fbcon_par;                /* fbcon use-only private area */
+
+	struct fb_image splash;
+
 	/* From here on everything is device dependent */
 	void *par;	
 };
diff -Nru a/include/linux/sysctl.h b/include/linux/sysctl.h
--- a/include/linux/sysctl.h	2005-02-13 14:35:45 +01:00
+++ b/include/linux/sysctl.h	2005-02-13 14:35:45 +01:00
@@ -135,6 +135,7 @@
 	KERN_HZ_TIMER=65,	/* int: hz timer on or off */
 	KERN_UNKNOWN_NMI_PANIC=66, /* int: unknown nmi panic flag */
 	KERN_BOOTLOADER_TYPE=67, /* int: boot loader type */
+	KERN_FBSPLASH=68,	/* string: path to fbsplash helper */
 };
 
 
diff -Nru a/kernel/panic.c b/kernel/panic.c
--- a/kernel/panic.c	2005-02-13 14:35:45 +01:00
+++ b/kernel/panic.c	2005-02-13 14:35:45 +01:00
@@ -18,6 +18,7 @@
 #include <linux/sysrq.h>
 #include <linux/interrupt.h>
 #include <linux/nmi.h>
+#include "../drivers/video/fbsplash.h"
 
 int panic_timeout;
 int panic_on_oops;
@@ -87,6 +88,8 @@
 		 * We can't use the "normal" timers since we just panicked..
 	 	 */
 		printk(KERN_EMERG "Rebooting in %d seconds..",panic_timeout);
+		fbsplash_verbose();
+		
 		for (i = 0; i < panic_timeout*1000; ) {
 			touch_nmi_watchdog();
 			i += panic_blink(i);
@@ -112,6 +115,8 @@
         disabled_wait(caller);
 #endif
 	local_irq_enable();
+	fbsplash_verbose();
+
 	for (i = 0;;) {
 		i += panic_blink(i);
 		mdelay(1);
diff -Nru a/kernel/power/console.c b/kernel/power/console.c
--- a/kernel/power/console.c	2005-02-13 14:35:45 +01:00
+++ b/kernel/power/console.c	2005-02-13 14:35:45 +01:00
@@ -9,6 +9,8 @@
 #include <linux/console.h>
 #include "power.h"
 
+#include "../drivers/video/fbsplash.h"
+
 static int new_loglevel = 10;
 static int orig_loglevel;
 #ifdef SUSPEND_CONSOLE
@@ -21,6 +23,9 @@
 	console_loglevel = new_loglevel;
 
 #ifdef SUSPEND_CONSOLE
+	if (fbsplash_issilent)
+		fbsplash_verbose();
+
 	acquire_console_sem();
 
 	orig_fgconsole = fg_console;
diff -Nru a/kernel/sysctl.c b/kernel/sysctl.c
--- a/kernel/sysctl.c	2005-02-13 14:35:45 +01:00
+++ b/kernel/sysctl.c	2005-02-13 14:35:45 +01:00
@@ -84,6 +84,9 @@
 #ifdef CONFIG_HOTPLUG
 extern char hotplug_path[];
 #endif
+#ifdef CONFIG_FB_SPLASH
+extern char fbsplash_path[];
+#endif
 #ifdef CONFIG_CHR_DEV_SG
 extern int sg_big_buff;
 #endif
@@ -393,6 +396,17 @@
 		.procname	= "hotplug",
 		.data		= &hotplug_path,
 		.maxlen		= HOTPLUG_PATH_LEN,
+		.mode		= 0644,
+		.proc_handler	= &proc_dostring,
+		.strategy	= &sysctl_string,
+	},
+#endif
+#ifdef CONFIG_FB_SPLASH
+	{
+		.ctl_name	= KERN_FBSPLASH,
+		.procname	= "fbsplash",
+		.data		= &fbsplash_path,
+		.maxlen		= KMOD_PATH_LEN,
 		.mode		= 0644,
 		.proc_handler	= &proc_dostring,
 		.strategy	= &sysctl_string,
